internal/gatewayapi/tls.go

```diff
@@ -9,53 +9,169 @@ import (
 	"crypto/x509"
 	"encoding/pem"
 	"fmt"
+	"strings"
+	"unicode/utf8"
 
 	corev1 "k8s.io/api/core/v1"
 )
 
 // validateTLSSecretData ensures the cert and key provided in a secret
 // is not malformed and can be properly parsed
-func validateTLSSecretData(secret *corev1.Secret) error {
-	certData := secret.Data[corev1.TLSCertKey]
+func validateTLSSecretsData(secrets []*corev1.Secret, domain string) error {
+	var publicKeyAlgorithm string
+	var certFQDN string
+	var parseErr error
+
+	pkaSecretSet := make(map[string]string)
+	for _, secret := range secrets {
+		certData := secret.Data[corev1.TLSCertKey]
+
+		certBlock, _ := pem.Decode(certData)
+		if certBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSCertKey)
+		}
+
+		cert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
+		}
+		publicKeyAlgorithm = cert.PublicKeyAlgorithm.String()
+
+		keyData := secret.Data[corev1.TLSPrivateKeyKey]
+
+		keyBlock, _ := pem.Decode(keyData)
+		if keyBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
+		}
+
+		if domain != "" {
+			err = verifyHostname(cert, domain)
+			if err != nil {
+				return fmt.Errorf("hostname %s does not match Common Name or DNS Names in the certificate %s", domain, corev1.TLSCertKey)
+			}
+		}
+		certFQDN = getFQDN(cert)
+		// Check whether the public key algorithm in the referenced secrets are unique.
+		if certFQDN, ok := pkaSecretSet[publicKeyAlgorithm]; ok {
```

Comment:
I think it should be possible to have multiple secrets/certs with the same `algo` and different CN/SAN
https://gateway.envoyproxy.io/v0.4.0/user/secure-gateways.html#sni-based-certificate-selection

Reply:
This will catch any secrets with the same algo and the same CN/SAN

Reply:
but you're only keying in the `algo` and not checking the CN/SAN
you probably want to do 
`if gotCertFQDN, ok := pkaSecretSet[publicKeyAlgorithm]; ok && certFQDN == gotCertFQDN {`
even this logic can be improved if `fqdn` is a []string
so it looks more like `if gotCertFQDN, ok := pkaSecretSet[publicKeyAlgorithm]; ok && len(intersection(certFQDNs,gotCertFQDNs)) > 0 {`

Reply:
Ah, you are right